### PR TITLE
Use api endpoint from vaults to upload/edit documents and tables

### LINKED ISSUE #6921: [Vaults UI] Add endpoints in vault to create/delete documents and tables
Existing endpoints: 

For documents :
create/edit = POST `/api/w/[wId]/data_sources/[name]/documents/[documentId]`
delete = DELETE `/api/w/[wId]/data_sources/[name]/documents/[documentId]`

For tables :
create/edit = POST `/api/w/[wId]/data_sources/[name]/tables/csv` (id passed in body)
delete = DELETE `/api/w/[wId]/data_sources/[name]/tables/[tableId]`
 
We need endpoints using vaults to correctly check the access permissions, ideally supporting GET/POST/PATCH/DELETE :
`/api/w/[wId]/vaults/[vId]/data_sources/[name]/document/[documentId]`, 
`/api/w/[wId]/vaults/[vId]/data_sources/[name]/tables/[tableId]`

### PR DESCRIPTION
**Pull request by tdraier**
fixes: https://github.com/dust-tt/dust/issues/6921

## Description

Add new endpoints in /vaults to create, edit and delete tables and documents. Use them instead of generic /datasources endpoints.

## Risk

Vault ui, can be rollbacked

## Deploy Plan

deploy front

### CODE REVIEW

File: front/pages/api/w/[wId]/data_sources/[name]/documents/[documentId]/index.ts [hunk 1]
```diff
@@ -38,9 +35,19 @@ async function handler(
   res: NextApiResponse<WithAPIErrorResponse<GetDocumentResponseBody>>,
   auth: Authenticator
 ): Promise<void> {
-  const plan = auth.getNonNullablePlan();
+  const { documentId, name } = req.query;
 
-  const dataSource = await getDataSource(auth, req.query.name as string);
+  if (typeof name !== "string" || typeof documentId !== "string") {
```
Comment (JulesBelveze) [inline front/pages/api/w/[wId]/data_sources/[name]/documents/[documentId]/index.ts:40]: Shouldn't we use io-ts here? Wondering why we sometimes use it and sometimes not

File: front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      res.status(204).end();
+      return;
+
+    default:
+      return apiError(req, res, {
+        status_code: 405,
+        api_error: {
+          type: "method_not_supported_error",
+          message:
+            "The method passed is not supported, GET, POST or DELETE is expected.",
```
Comment (JulesBelveze) [inline front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts:199]: `nit`: we also support PATCH

File: front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/index.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        document: upsertResult.value.document,
+      });
+      return;
+    default:
+      return apiError(req, res, {
+        status_code: 405,
+        api_error: {
+          type: "method_not_supported_error",
+          message:
+            "The method passed is not supported, GET, POST or DELETE is expected.",
```
Comment (JulesBelveze) [inline front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/index.ts:145]: `nit`: we only support `POST` here

**Review by JulesBelveze — COMMENTED** [review]
Looks legit to me 💪🏼  let's wait for a second pair of 👁️ as this is a quite important piece of code.

Note: I would try to converge on the use of io-ts though, let's not rewrite all the existing code but we should probably move to io-ts when adding new endpoints

File: front/pages/api/w/[wId]/data_sources/[name]/documents/[documentId]/index.ts (hunk 1 above)
Comment (tdraier) [reply front/pages/api/w/[wId]/data_sources/[name]/documents/[documentId]/index.ts:40]: We're using iots for body parsing, here it's path parameters - but we could definitely use iots even if it's only simple strings.

File: front/lib/api/data_sources.ts [hunk 4]
```diff
@@ -211,3 +227,247 @@ export async function augmentDataSourceWithConnectorDetails(
...
+
+  const fullText = sectionFullText(generatedSection);
+
+  const coreAPI = new CoreAPI(apiConfig.getCoreAPIConfig(), logger);
+  const plan = auth.getNonNullablePlan();
+  // Enforce plan limits: DataSource documents count.
+  // We only load the number of documents if the limit is not -1 (unlimited).
+  // the `getDataSourceDocuments` query involves a SELECT COUNT(*) in the DB that is not
+  // optimized, so we avoid it for large workspaces if we know we're unlimited anyway
+  if (plan.limits.dataSources.documents.count != -1) {
```
Comment (flvndvd) [inline front/lib/api/data_sources.ts:323]:
```suggestion
  if (plan.limits.dataSources.documents.count !== -1) {
```

File: front/lib/api/data_sources.ts [hunk 5]
```diff
@@ -211,3 +227,247 @@ export async function augmentDataSourceWithConnectorDetails(
...
+          `Contact support@dust.tt if you want to increase it.`
+      )
+    );
+  }
+
+  // Data source operations are performed with our credentials.
+  const credentials = dustManagedCredentials();
+
+  // Create document with the Dust internal API.
+  const upsertRes = await coreAPI.upsertDataSourceDocument({
```
Comment (flvndvd) [inline front/lib/api/data_sources.ts:373]: @philipperolet is it expected that uploading a document on a static data source from the app does not go through. the upsert queue?

File: front/lib/api/data_sources.ts [hunk 6]
```diff
@@ -211,3 +227,247 @@ export async function augmentDataSourceWithConnectorDetails(
...
+      },
+    });
+    if (enqueueRes.isErr()) {
+      return enqueueRes;
+    }
+
+    return new Ok(undefined);
+  }
+
+  const tableRes = await upsertTableFromCsv({
```
Comment (flvndvd) [inline front/lib/api/data_sources.ts:459]: @philipperolet is it expected that uploading a document on a static data source from the app does not go through. the upsert queue?

File: front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+      api_error: {
+        type: "data_source_not_found",
+        message: "The data source you requested was not found.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "PATCH":
+      if (!auth.isBuilder() && !dataSource.canWrite(auth)) {
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts:73]: Shouldn't we include `isBuilder` in the `DataSourceResource.canWrite()`?

File: front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        return apiError(req, res, {
+          status_code: 403,
+          api_error: {
+            type: "data_source_auth_error",
+            message: "You are not allowed to update data in this data source.",
+          },
+        });
+      }
+
+      if (dataSource.connectorId) {
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts:83]: `nit`

We can use `isWebsite` and `isManaged`?

File: front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+          },
+        });
+      }
+
+      const coreAPI = new CoreAPI(apiConfig.getCoreAPIConfig(), logger);
+
+      const deleteRes = await coreAPI.deleteDataSourceDocument({
+        projectId: dataSource.dustAPIProjectId,
+        dataSourceId: dataSource.dustAPIDataSourceId,
+        documentId: req.query.documentId as string,
```
Comment (flvndvd) [inline front/pages/api/w/[wId]/vaults/[vId]/data_sources/[dsId]/documents/[documentId]/index.ts:176]:
```suggestion
        documentId,
```

File: types/src/front/api_handlers/public/data_sources.ts [hunk 10]
```diff
@@ -35,6 +39,43 @@ export const PostDataSourceDocumentRequestBodySchema = t.type({
...
+
 export type PostDataSourceDocumentRequestBody = t.TypeOf<
   typeof PostDataSourceDocumentRequestBodySchema
 >;
+
+export const PatchDataSourceTableRequestBodySchema = t.type({
+  name: t.string,
+  description: t.string,
+  timestamp: t.union([t.number, t.undefined, t.null]),
+  tags: t.union([t.array(t.string), t.undefined, t.null]),
```
Comment (flvndvd) [inline types/src/front/api_handlers/public/data_sources.ts:61]: It feels very weird to have all those nullable options 🤔

**Review by flvndvd — APPROVED** [review]
LGTM

File: front/components/data_source/DocumentUploadOrEditModal.tsx [hunk 11]
```diff
@@ -142,11 +142,19 @@ export function DocumentUploadOrEditModal({
...
+        sendNotification({
+          type: "success",
+          title: "Document successfully updated",
+          description: `Your document ${documentId} was successfully updated`,
+        });
+      } else {
+        sendNotification({
+          type: "success",
+          title: "Document successfully added",
+          description: `Your document ${documentId} was successfully added`,
```
[the remark this answers is not in the document]
Comment (tdraier) [reply front/components/data_source/DocumentUploadOrEditModal.tsx]: fixed by @JulesBelveze elsewhere

File: front/lib/api/data_sources.ts (hunk 6 above)
Comment (tdraier) [reply front/lib/api/data_sources.ts:459]: for table we have an async param which allows to put it in the queue, but it is (was) set to false.

File: types/src/front/api_handlers/public/data_sources.ts (hunk 10 above)
Comment (tdraier) [reply types/src/front/api_handlers/public/data_sources.ts:61]: Yes, very weird. I think we should keep only value or undefined, but seems decoder does not handles all cases properly, and we pass null in form body in some cases. Some cleanup needed here.

**Pull request merged**